### PR TITLE
✨ Add ca-cert flag to supply a default ca certificate for each requests

### LINKED ISSUE #1439: Add `--ca-cert` flag to supply default ca certificates in requests.
/kind feature

**Describe the solution you'd like**

After discussions on #1434 it has been stated that adding a `--ca-cert` flag could be a good idea.

Actually using CAPO, users have to add the `cacert` key in each secrets for each clusters. It can becomes complex in the case where many clusters are created by many tenants on the same (or public) openstack endpoints. 

Adding the `--ca-cert` will allow infrastructure operators to set default ca certificates to supply to each requests when no `cacert` key is provided in the cloud-config secret alongside the `clouds.yaml`. 

**Anything else you would like to add:**

You can follow the whole discussions in #1434 to get more context and most of our reflexions. 
This solution looks better than solution described in #1433

### PR DESCRIPTION
**Pull request by alexandrevilain**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1439

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by jichenjc** [comment]
/lgtm

follow up we may guide user how to utilize this flag, e.g some doc update etc..

**Comment by alexandrevilain** [comment]
Thanks @jichenjc 

For sure, I'll create a documentation PR about ca-certificate behavior as soon as possible :)

**Comment by mdbooth** [comment]
This is great, thanks! Just trying to think of a way to test it; it doesn't fit into any of our existing test patterns. Any ideas?

**Comment by jichenjc** [comment]
I think we don't have https enabled in CI? maybe we can create devstack with https enabled, and find one of full test (not for every PR) to run https , thus we first cover https test along with this PR (with the flag)?

**Comment by alexandrevilain** [comment]
So you want me to implement e2e test for this feature ?
In a follow up PR alongside documentation ?

**Comment by jichenjc** [comment]
>>>In a follow up PR alongside documentation ?

I think we can use follow up PR , not a must have for this PR

File: main.go [hunk 1]
```diff
@@ -161,6 +164,15 @@ func main() {
 		setupLog.Error(err, "unable to get kubeconfig")
 	}
 
+	var cacerts []byte
+	if cacertsPath != "" {
```
Comment (seanschneeweiss) [inline main.go:168]:
```suggestion
	var caCerts []byte
	if caCertsPath != "" {
```

File: main.go [hunk 2]
```diff
@@ -133,6 +134,8 @@ func InitFlags(fs *pflag.FlagSet) {
 
 	fs.StringVar(&lbProvider, "lb-provider", "amphora",
 		"The name of the load balancer provider (amphora or ovn) to use (defaults to amphora).")
+
+	fs.StringVar(&cacertsPath, "ca-certs", "", "The path to a ca certificate file to supply as default cacert for each requests.")
```
Comment (seanschneeweiss) [inline main.go:138]: Might be good to document here, if the secret takes precedence over the default path provided here.
```suggestion
	fs.StringVar(&cacertsPath, "ca-certs", "", "The path to a PEM-encoded CA Certificate file to supply as default for each request.")
```

File: pkg/cloud/services/provider/provider.go [hunk 3]
```diff
@@ -67,6 +72,11 @@ func NewClientFromCluster(ctx context.Context, ctrlClient client.Client, openSta
 			return nil, nil, "", err
 		}
 	}
+
+	if caCert == nil && defaultCACert != nil {
```
Comment (seanschneeweiss) [inline pkg/cloud/services/provider/provider.go:76]: Considerung `defaultCACert == nil` and  `caCert == nil`, it shouldn't matter if we set `caCert = nil` or leave it `nil`.
```suggestion
	if caCert == nil {
```

**Review by seanschneeweiss — COMMENTED** [review]
Thank you for that contribution. Following the discussion of #1434, providing a default CaCert rather than having to set it in each cluster's secret seems reasonable.

**Comment by alexandrevilain** [comment]
Thanks for the review @seanschneeweiss !
I applied all your comments.

**Comment by alexandrevilain** [comment]
/retest

I don't understand why tests are failing ...

**Comment by alexandrevilain** [comment]
e2e are ok since my rebase 🎉

**Comment by mdbooth** [comment]
/approve

Thanks! This should just need lgtm from somebody now.

**Pull request merged**